### ISSUE TITLE
in keyword can't be used on a narrowed down expression with a nullable generic constraint

### ISSUE THREAD

**Issue by Andarist** [microsoft/TypeScript#51501]
# Bug Report

### 🔎 Search Terms

in keyword, narrow, nullable, NonNullable, generic constraint

### 🕗 Version & Regression Information

- This is related to the upcoming TS 4.9 release that comes with [the improved support for the in keyword](https://github.com/microsoft/TypeScript/pull/50666)

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=5.0.0-dev.20221112#code/KYDwDg9gTgLgBAMwK4DsDGMCWEV0wZwAkAVAWQBliBDAIwBtgAeYuUGYFAE3zghoCtgGOAB84KJHToA+ABQxaDAFxxiAShU0IEBlVwBvAFBw4UYDCRRcAQmsL6wOADIncAOQALGAFs6bvLj2DADchgC+hoagkLCIqBjYuAQkFNQOAEzyisAqfILCYhJSGnBaOsB6cEYmZhZWcLZBji7uXr7+mIHZoWFAA)

### 💻 Code

```ts
export function isHTMLTable<T extends object | null>(table: T): boolean {
  return !!table && 'html' in table; // Type 'NonNullable<T>' may represent a primitive value, which is not permitted as the right operand of the 'in' operator.(2638)
}

export function isHTMLTable2(table: object | null): boolean {
  return !!table && 'html' in table; // no error
}
```

### 🙁 Actual behavior

The generic version of this function is reporting an error on the `in` operator but the non-generic one isn't

### 🙂 Expected behavior

I believe that both of them should not error.

_This was originally reported by @jcalz [here](https://github.com/microsoft/TypeScript/issues/51007#issuecomment-1312093648)_

**Comment by wegry** [comment]
As a potential workaround, [there is this stackoverflow answer](https://stackoverflow.com/a/74413642/1924257).

**Comment by HolgerJeromin** [comment]
Could/should this be cherry picked to 4.9 branch?
cc @RyanCavanaugh 
makes removing away from `suppressImplicitAnyIndexErrors` a bit easier.
